src/zenml/integrations/modal/orchestrators/modal_orchestrator.py

```diff
@@ -0,0 +1,206 @@
+
+        Returns:
+            A `StackValidator` instance.
+        """
+        validator: StackValidator = create_modal_stack_validator()
+        return validator
+
+    def get_orchestrator_run_id(self) -> str:
+        """Returns the active orchestrator run id.
+
+        Raises:
+            RuntimeError: If the environment variable specifying the run id
+                is not set.
+
+        Returns:
+            The orchestrator run id.
+        """
+        try:
+            return os.environ[ENV_ZENML_MODAL_ORCHESTRATOR_RUN_ID]
+        except KeyError:
+            raise RuntimeError(
+                "Unable to read run id from environment variable "
+                f"{ENV_ZENML_MODAL_ORCHESTRATOR_RUN_ID}."
+            )
+
+    def prepare_or_run_pipeline(
+        self,
+        deployment: "PipelineDeploymentResponse",
+        stack: "Stack",
+        environment: Dict[str, str],
+        placeholder_run: Optional["PipelineRunResponse"] = None,
+    ) -> Optional[Iterator[Dict[str, MetadataType]]]:
+        """Runs the complete pipeline using Modal sandboxes.
+
+        Args:
+            deployment: The pipeline deployment to prepare or run.
+            stack: The stack the pipeline will run on.
+            environment: Environment variables to set in the orchestration
+                environment.
+            placeholder_run: An optional placeholder run for the deployment.
+
+        Raises:
+            Exception: If pipeline execution fails.
+
+        Returns:
+            None if the pipeline is executed synchronously, otherwise an iterator of metadata dictionaries.
+        """
+        if deployment.schedule:
+            logger.warning(
+                "Modal Orchestrator currently does not support the "
+                "use of schedules. The `schedule` will be ignored "
+                "and the pipeline will be run immediately."
+            )
+
+        # Setup Modal authentication
+        self._setup_modal_client()
+
+        # Generate orchestrator run ID and include pipeline run ID for isolation
+        orchestrator_run_id = str(uuid4())
+        environment[ENV_ZENML_MODAL_ORCHESTRATOR_RUN_ID] = orchestrator_run_id
+
+        # Pass pipeline run ID for proper isolation
+        if placeholder_run:
+            environment["ZENML_PIPELINE_RUN_ID"] = str(placeholder_run.id)
+            logger.debug(f"Pipeline run ID: {placeholder_run.id}")
+
+        logger.debug(f"Orchestrator run ID: {orchestrator_run_id}")
+
+        # Get settings from pipeline configuration
+        settings = cast(
+            "ModalOrchestratorSettings", self.get_settings(deployment)
+        )
+
+        # Check execution mode
+        execution_mode = getattr(
+            settings, "execution_mode", ModalExecutionMode.PIPELINE
+        )
+        logger.info(
+            f"🚀 Executing pipeline with Modal ({execution_mode.lower()} mode)"
+        )
+
```

Comment:
If the execution mode is pipeline, we do not allow per-step images (as they cannot be used). I suggest you overwrite the `get_docker_builds(...)` method as follows:
- Call the super implementation
- Check if there is any build configuration in the result that has `key="orchestrator"` and a non-None `step_name`. In that case, raise an error.